### PR TITLE
Some improvements to the resourceRequest controller.

### PR DESCRIPTION
**Pull request by giuse2596**
# Description

- [x] Add Owns() and filter options to For() in SetuoWithManager function.
- [x] Create a new ForeignCluster if it does not exists.
- [x] Fix Node Informers Running check substituting Node.Status.Phase (deprecated and never used) with Node.Status.Conditions lookup.
- [x] Fix klog import in Broadcaster.

### CODE REVIEW

File: internal/resource-request-operator/resourceRequest_controller.go [hunk 1]
```diff
@@ -41,6 +44,25 @@ func (r *ResourceRequestReconciler) Reconcile(ctx context.Context, req ctrl.Requ
...
+	err = r.List(ctx, foreignClusterList, client.MatchingLabels{
+		discovery.ClusterIDLabel: resourceRequest.Spec.ClusterIdentity.ClusterID,
+	})
+
+	if err != nil {
+		klog.Errorf("%s -> unable to List foreignCluster with clusterID %s: %s", r.ClusterID, resourceRequest.Spec.ClusterIdentity.ClusterID, err)
+		return ctrl.Result{}, nil
+	}
+
+	if len(foreignClusterList.Items) == 0 {
```
Comment (palexster) [inline internal/resource-request-operator/resourceRequest_controller.go:57]: Does deserve a bit more commentary? Is this related to the Incoming peer?

File: internal/resource-request-operator/utils.go [hunk 2]
```diff
@@ -51,3 +52,29 @@ func (r *ResourceRequestReconciler) generateResourceOffer(request *discoveryv1al
 	klog.Infof("%s -> %s Offer: %s", r.ClusterID, op, offer.ObjectMeta.Name)
 	return nil
 }
+
+func (r *ResourceRequestReconciler) createForeignCluster(resourceRequest *discoveryv1alpha1.ResourceRequest) error {
```
Comment (palexster) [inline internal/resource-request-operator/utils.go:56]: Ditto.

File: internal/resource-request-operator/broadcaster.go [hunk 3]
```diff
@@ -268,3 +268,12 @@ func extractPodResources(pod *corev1.Pod) corev1.ResourceList {
 	resourcesToExtract, _ := resourcehelper.PodRequestsAndLimits(pod)
 	return resourcesToExtract
 }
+
+func isNodeRunning(node *corev1.Node) bool {
```
Comment (giorio94) [inline internal/resource-request-operator/broadcaster.go:272]:
```suggestion
func isNodeReady(node *corev1.Node) bool {
```

Since this function may be of use also in other situations (e.g. VK node controller), I would propose to move this outside of this specific package

File: internal/resource-request-operator/utils.go [hunk 4]
```diff
@@ -51,3 +52,29 @@ func (r *ResourceRequestReconciler) generateResourceOffer(request *discoveryv1al
...
+		Spec: discoveryv1alpha1.ForeignClusterSpec{
+			ClusterIdentity: resourceRequest.Spec.ClusterIdentity,
+			Namespace:       resourceRequest.Namespace,
+			Join:            false,
+			DiscoveryType:   discovery.IncomingPeeringDiscovery,
+			AuthURL:         resourceRequest.Spec.AuthURL,
+		},
+	}
+
+	err := r.Create(context.Background(), foreignCluster)
```
Comment (giorio94) [inline internal/resource-request-operator/utils.go:74]: It seems to me at this point the resourcerequest should become the owner of this foreigncluster to ensure eventual deletion. @aleoli Is that correct?

File: internal/resource-request-operator/utils.go [hunk 5]
```diff
@@ -51,3 +52,29 @@ func (r *ResourceRequestReconciler) generateResourceOffer(request *discoveryv1al
...
+			},
+		},
+		Spec: discoveryv1alpha1.ForeignClusterSpec{
+			ClusterIdentity: resourceRequest.Spec.ClusterIdentity,
+			Namespace:       resourceRequest.Namespace,
+			Join:            false,
+			DiscoveryType:   discovery.IncomingPeeringDiscovery,
+			AuthURL:         resourceRequest.Spec.AuthURL,
+		},
+	}
```
Comment (giorio94) [inline internal/resource-request-operator/utils.go:72]: @aleoli Is it sufficient to create this foreigncluster or its definition should also be enforced in case of modifications?

**Review by giorio94 — COMMENTED** [review]
A couple of comments inline.
I definitely agree with @palexster that more comments are needed to clarify the workflow

File: internal/resource-request-operator/utils.go (hunk 4 above)
Comment (aleoli) [reply internal/resource-request-operator/utils.go:74]: No, it can't, it is a namespaced resource while the ForeignCluster is cluster-wide. The deletion logic is inside the ForeignCluster Operator.
It has to be set the opposite, the ResourceRequest ahs to be owned by the ForeignClsuter to ensure FC Status updates on incoming peering chages.

File: internal/resource-request-operator/utils.go (hunk 5 above)
Comment (aleoli) [reply internal/resource-request-operator/utils.go:72]: It should be enforced, but only if the discovery type is still IncomingPeering, since it can be discovered in other ways (i.e. mDNS, DNS) that can overwrite this method and change its content.
And then we have to take care and not enforce the join flag, to preserve user chooses.

Comment (giorio94) [reply internal/resource-request-operator/utils.go:72]: Got it. As agreed, for the moment it is probably better not to enforce it to avoid problems, since all fields should be anyway immutable (at least conceptually)

File: internal/resource-request-operator/utils.go (hunk 4 above)
Comment (giorio94) [reply internal/resource-request-operator/utils.go:74]: Got it. Indeed, as discussed, it should be configured the reverse owner-reference (the resource request is owned by the foreigncluster), to ensure the status is correctly reconciled. Once those two operations are performed, it is probably better to return and create the resourceoffer in the next reconcile, to prevent self-created conflicts

File: internal/resource-request-operator/resourceRequest_controller.go [hunk 6]
```diff
@@ -41,7 +44,29 @@ func (r *ResourceRequestReconciler) Reconcile(ctx context.Context, req ctrl.Requ
...
+		// if does not exist any foreignCluster which satisfies above conditions create a new one.
```
Comment (giorio94) [inline internal/resource-request-operator/resourceRequest_controller.go:60]: IMHO, this comment has limited value, since it mostly repeats what is obvious from the code. In comments you should mostly elaborate on high level logic, why it is needed and why certain conditions can occur (e.g. why the foreigncluster may/may not exist), without repeating low level details.

File: internal/resource-request-operator/resourceRequest_controller.go [hunk 7]
```diff
@@ -41,7 +44,29 @@ func (r *ResourceRequestReconciler) Reconcile(ctx context.Context, req ctrl.Requ
 		return ctrl.Result{}, nil
 	}
 
-	offerErr := r.generateResourceOffer(&resourceRequest)
+	// check if a foreignCluster with an incoming peering
```
Comment (giorio94) [inline internal/resource-request-operator/resourceRequest_controller.go:47]: The comment mentions the incoming peering, but it seems to me this is not checked anywhere (correctly AFAIU). In that case the comment should be rephrased.

File: pkg/utils/nodeUtils.go [hunk 8]
```diff
@@ -0,0 +1,13 @@
+package utils
+
+import corev1 "k8s.io/api/core/v1"
+
+// IsNodeReady return true if the passed node has the NodeReady condition = True, false otherwise.
```
Comment (giorio94) [inline pkg/utils/nodeUtils.go:5]:
```suggestion
// IsNodeReady returns true if the passed node has the NodeReady condition equal to True, false otherwise.
```

**Review by palexster — APPROVED** [review]

**Review by giorio94 — APPROVED** [review]
